### PR TITLE
[Fix] Allow connection of a noban banned peer

### PR DESCRIPTION
**Pull request by NicolasDorier**
Reported by @MarcoFalke on https://github.com/bitcoin/bitcoin/pull/16248#discussion_r314026195

The bug would mean that if the peer connecting to you is banned, but whitelisted without specific permissions, it would not be able to connect to the node.

The solution is just to move the same line below.

### CODE REVIEW

**Comment by gapeman** [comment]
Is this safe to do? What are the implications of connecting to a peer that would otherwise be banned?

**Comment by Sjors** [comment]
@gapeman afaik that was the behavior before #16248.

@NicolasDorier would be nice to have a test for this.

**Comment by NicolasDorier** [comment]
@gapeman if your config specifically say that you don't want to ban a peer, then it should not be banned.

As @Sjors says, I just want to keep current behavior working as before.

I can take a look to see if I can make a test for that.

**Comment by NicolasDorier** [comment]
I refactored it to properly scope some variables. Writing a test so such regression can't happen.

**Comment by NicolasDorier** [comment]
Fixed the bug, and added an anti regression test for `setban`

**Comment by naumenkogs** [comment]
utACK a0d2e2b

File: test/functional/rpc_setban.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
+# Copyright (c) 2015-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the setban rpc call."""
+
+from test_framework.test_node import ErrorMatch
```
Comment (MarcoFalke) [inline test/functional/rpc_setban.py]: test/functional/rpc_setban.py:7:1: F401 'test_framework.test_node.ErrorMatch' imported but unused

test/functional/rpc_setban.py:39:78: F821 undefined name 'p'

test/functional/rpc_setban.py:41:75: F821 undefined name 'p'

File: test/functional/rpc_setban.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        self.restart_node(1, [])
+        connect_nodes(self.nodes[0], 1)
+        peerinfo = self.nodes[1].getpeerinfo()
+        assert_equal(0, len(peerinfo))
+        # Node 1 should be able to reconnect if it has noban permission
+        self.restart_node(1, ['-whitelist=127.0.0.1'])
+        connect_nodes(self.nodes[0], 1)
+        self.assert_has_permission('noban', True)
+
+    def assert_has_permission(self, permission, contains):
```
Comment (Sjors) [inline test/functional/rpc_setban.py]: Maybe pass node in directly and split into `assert_has_permission(self, node, permission)` and `assert_has_no_permission(self, node, permission)`

**Review by Sjors — COMMENTED** [review]
Travis failed due to lint issue. Maybe rename `rpc_setban.py` to `p2p_whitelist.py`?

File: test/functional/rpc_setban.py (hunk 2 above)
Comment (NicolasDorier) [reply test/functional/rpc_setban.py]: what about `assert_equal(True, self.hasPermision(...))` ?

That said, I don't think it should pass the nodes to this method. It is confusing (there is two nodes: the node that is permissioned, and the node on which we want to check permissions).

Comment (Sjors) [reply test/functional/rpc_setban.py]: `assert(self.hasPermision(...))` is even simpler.

You could pass both nodes. Or you could document at the top of the test file which hardcoded nodes do what.

Comment (promag) [reply test/functional/rpc_setban.py]: Sorry but I think you could just inline this - it's longer as it is with the `contains` arg.

**Review by promag — COMMENTED** [review]
ACK

**Comment by NicolasDorier** [comment]
As @promag suggested, I instead just inlined the permission checked, it is actually more readable.

I also check in the logs that we really prevent connection because of the ban. I also added a test to `setban x remove`.

**Comment by NicolasDorier** [comment]
@Sjors `rpc_setban` is the right name, I really want to test setban, so I added it in the whitelist of authorized term.

**Comment by Sjors** [comment]
Agree inline is more clear. utACK d117f45

**Comment by MarcoFalke** [comment]
ACK d117f4541d4717e83c9396273e92960723622030

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK d117f4541d4717e83c9396273e92960723622030
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUin2gv9GZKDSUbn62676zj0G6Igb8tzgC2M1Sf2KyEuIR+heCoVYGlyp3rXDbuc
22c5dyDXDnktsyMR3mwo3KLzT/MBY/wc8nDgdzkRUzWt6j+/q8Ex8NTnfbgyGXzr
vBE3bJOfA1y3DCCzpSar4uU/sPUAK+oJ3qF1+62rijDYpC0Jh1UaYMy7AdYQdYQ/
kyO3jorAiwIk0sE7SzJDq3rE7wlHzQ4LCZauCnVft/Y1rOY1IkmYfke3uapR4EVr
Zw7D40jY6RsEDLNh/1odaRqjm01v1uJ5d7mFnqOFCKA7RwdS2NQ0ElbfQ3fAUowm
V7Oola76XE8OvGsj9y+qOwBkX5iMgeNQANb+Pf6poB0ANDh0AKpIkxjSIGGbhKi2
XN9KGRaHLpNBeU69lQDrvsXEzyfl4kSLyO9/k7+qZAtgUTqyLDhi9beqS9Ta0OKF
X7RHymgGIn1zvTOY4WSoPiHIF5ceVDu9S5r/fxr9DHmwPWWEzOE9ztfcSiQw68tY
HRrWUvc+
=qI3j
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `74a14b73630691da8ffa64e5b597d7fd7246787327bba54579da5c03c881ec01  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010874a14b73630691da8ffa64e5b597d7fd7246787327bba54579da5c03c881ec01f01043219eeffe1ae9a1026ee48a3dd50b9308fff01010030cd40044e4372cb6d2446292049608f020f11ddc7314ce1f42373e025745cf37f58ab5434f875df973797c6bc7f768a88a08f1045d56b8e1f0088dbe7bdd983eab1a0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01019f975197112f255b780d679cad9efc208f1045d56b8e0f00883cdfc66b55a8b5b0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0102e30f02eece966fef2a0c783189e008908f1045d56b8e0f008182705ebec4d87280083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df0105f28dc7979c351e16ded943d6a738b2e08f1045d56b8e0f008952e9f3fb11641920083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

**Pull request merged**